libs/residence-history/src/lib/computeCountryResidence.spec.ts

```diff
@@ -0,0 +1,130 @@
+import { computeCountryResidence } from './computeCountryResidence'
+import { DAY } from './constants'
+
+jest
+  .useFakeTimers('modern')
+  .setSystemTime(new Date('2021-09-18T00:00:00').getTime())
+
+const d = new Date().getTime()
+
+// helper to return a date object with subtracted days from "today"
+const makeDate = (addDays: number): Date => new Date(d + addDays * DAY)
+
+describe('computeCountryResidence()', () => {
+  it('should work with a single entry older than a year', async () => {
+    const report = computeCountryResidence([
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-500),
+      },
+    ])
+
+    expect(report).toMatchObject({
+      IS: 365,
+    })
+  })
+
+  it('should work with a single entry less than a year, but more than 185 days', async () => {
+    const report = computeCountryResidence([
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-200),
+      },
+    ])
+
+    expect(report).toMatchObject({
+      IS: 200,
+    })
+  })
+
+  it('should work with a single entry more than a year, where residence was abroad', async () => {
+    const report = computeCountryResidence([
+      {
+        address: {},
+        country: 'US',
+        dateOfChange: makeDate(-500),
+      },
+    ])
+
+    expect(report).toMatchObject({
+      US: 365,
+    })
+  })
+
+  it('should work with mixed residences', async () => {
+    const report = computeCountryResidence([
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-2),
+      },
+      {
+        address: {},
+        country: 'US',
+        dateOfChange: makeDate(-4),
+      },
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-6),
+      },
+    ])
+
+    expect(report).toMatchObject({
+      IS: 4,
+      US: 2,
+    })
+  })
+
+  it('should work with ancient residences, even though most recent is abroad', async () => {
+    const report = computeCountryResidence([
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-1500),
+      },
+      {
+        address: {},
+        country: 'US',
+        dateOfChange: makeDate(-100),
+      },
+    ])
+
+    expect(report).toMatchObject({
+      IS: 265,
+      US: 100,
+    })
+  })
+
+  it('should work with dates even when they are returned out of order', async () => {
+    const report = computeCountryResidence([
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-1500),
+      },
+      {
+        address: {},
+        country: 'US',
+        dateOfChange: makeDate(-20),
+      },
+      {
+        address: {},
+        country: 'US',
+        dateOfChange: makeDate(-100),
+      },
+      {
+        address: {},
+        country: 'IS',
+        dateOfChange: makeDate(-300),
+      },
+    ])
+
+    expect(report).toMatchObject({
+      IS: 265,
+      US: 100,
+    })
+  })
```

Comment:
It does not, the date calculations are pretty complex but I added some console.logs you can see here hope they help!

US 2021-08-29T00:00:00.000Z
CURRENT =  Math.round(( 2021-09-18T00:00:00.000Z  - 2021-08-29T00:00:00.000Z ) / DAY 86400000 )
const period = Math.round(( 2021-09-18T00:00:00.000Z - 2021-08-29T00:00:00.000Z ) / DAY 86400000 )
timeByCountry[country] = (timeByCountry[country] 20  || 0) + period 20
lastTime = current = 2021-08-29T00:00:00.000Z
US 2021-06-10T00:00:00.000Z
CURRENT =  Math.round(( 2021-08-29T00:00:00.000Z  - 2021-06-10T00:00:00.000Z ) / DAY 86400000 )
const period = Math.round(( 2021-08-29T00:00:00.000Z - 2021-06-10T00:00:00.000Z ) / DAY 86400000 )
timeByCountry[country] = (timeByCountry[country] 100  || 0) + period 80
lastTime = current = 2021-06-10T00:00:00.000Z
IS 2020-11-22T00:00:00.000Z
CURRENT =  Math.round(( 2021-06-10T00:00:00.000Z  - 2020-11-22T00:00:00.000Z ) / DAY 86400000 )
const period = Math.round(( 2021-06-10T00:00:00.000Z - 2020-11-22T00:00:00.000Z ) / DAY 86400000 )
timeByCountry[country] = (timeByCountry[country] 200  || 0) + period 200
lastTime = current = 2020-11-22T00:00:00.000Z
IS 2017-08-10T00:00:00.000Z
CURRENT =  Math.round(( 2020-11-22T00:00:00.000Z  - 2020-09-18T00:00:00.000Z ) / DAY 86400000 )
const period = Math.round(( 2020-11-22T00:00:00.000Z - 2020-09-18T00:00:00.000Z ) / DAY 86400000 )
timeByCountry[country] = (timeByCountry[country] 265  || 0) + period 65
lastTime = current = 2020-09-18T00:00:00.000Z
{ US: 100, IS: 265 }

Reply:
I think I was a bit too quick with the Approval trigger. I still think the `should work with dates even when they are returned out of order` test case is invalid. If you print the current country under consideration I expect it to show `IS -> IS -> US -> US`. The test cases passes because `IS(-1500) -> US(-100)` gives  `{ US: 100, IS: 265 }`.